examples/dynamic-bridge-app/linux/include/DataModel.h

```diff
@@ -0,0 +1,218 @@
+#pragma once
+
+#include <app/AttributeAccessInterface.h>
+#include <app/data-model/Encode.h>
+#include <app/data-model/Decode.h>
+#include <app/data-model/Nullable.h>
+
+#include <vector>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+template<typename X>
+class IsList
+{
+    typedef char yes;
+    typedef long no;
+
+    template <typename U> static yes test(decltype(&U::size)) ;
+    template <typename U> static no test(...);
+
+public:
+    static constexpr bool value = sizeof(test<X>(0)) == sizeof(yes);
+};
+
+template<typename X>
+struct IsOptionalOrNullable { static constexpr bool value = false; };
+template<typename X>
+struct IsOptionalOrNullable<Optional<X>> { static constexpr bool value = true; };
+template<typename X>
+struct IsOptionalOrNullable<Nullable<X>> { static constexpr bool value = true; };
+
+static_assert(IsList<std::vector<unsigned char>>::value);
+
+template<typename X, std::enable_if_t<!IsOptionalOrNullable<std::decay_t<X>>::value && !IsList<std::decay_t<X>>::value, bool> = true>
+CHIP_ERROR Encode(const ConcreteReadAttributePath&, AttributeValueEncoder & aEncoder, const X& x)
+{
+    return aEncoder.Encode(x);
+}
+
+/*
+ * @brief
+ * Lists that are string-like should be encoded as char/byte spans.
+ */
+template<typename X, std::enable_if_t<IsList<std::decay_t<X>>::value && sizeof(std::decay_t<typename X::pointer>) == sizeof(char), bool> = true>
+CHIP_ERROR Encode(const ConcreteReadAttributePath& aPath, AttributeValueEncoder & aEncoder, const X& x)
+{
+    return aEncoder.Encode(Span<std::decay_t<typename X::pointer>>(x.data(), x.size()));
+}
+
+
+/*
+ * @brief
+ *
+ * If an item is requested from a list, encode just that single item, or the entire list otherwise.
+ *
+ * The object must satisfy the following constraints
+ * size() must return an integer
+ * begin() must return a type conforming to LegacyRandomAccessIterator
```

Comment:
Does it have to be a `LegacyRandomAccessIterator`? Couldn't we relax this to be a `LegacyForwardIterator` to then permit use of `std::list` (and other similar containers) if needed as well?

Reply:
Yes if we are willing to repeatedly increment the iterator instead of assuming we can use indexing

Reply:
Changed to use std::advance